### PR TITLE
CDAP-1665: Refactor Store to throw NotFoundException, etc. where appropriate instead of returning null

### PR DESCRIPTION
**Pull request by jwang47**
Based on https://github.com/caskdata/cdap/pull/1484 (merged) and https://github.com/caskdata/cdap/pull/1869

### CODE REVIEW

File: cdap-app-fabric/src/main/java/co/cask/cdap/app/store/Store.java [hunk 1]
```diff
@@ -119,7 +121,7 @@
    * @param appArchiveLocation location of the deployed app archive
    */
   void addApplication(Id.Application id,
-                      ApplicationSpecification specification, Location appArchiveLocation);
+                      ApplicationSpecification specification, Location appArchiveLocation) throws AlreadyExistsException, NotFoundException;
```
Comment (poornachandra) [inline cdap-app-fabric/src/main/java/co/cask/cdap/app/store/Store.java]: Line is longer than 120 chars

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AppLifecycleHttpHandler.java [hunk 2]
```diff
@@ -400,40 +376,28 @@ public void deleteAdapter(HttpRequest request, HttpResponder responder,
...
+    throws SchedulerException, NotFoundException, AlreadyExistsException, IOException {
 
-    try {
-      if (!namespaceAdmin.hasNamespace(Id.Namespace.from(namespaceId))) {
-        responder.sendString(HttpResponseStatus.NOT_FOUND,
-                             String.format("Create adapter failed - namespace '%s' does not exist.", namespaceId));
-        return;
-      }
+    if (!namespaceAdmin.hasNamespace(Id.Namespace.from(namespaceId))) {
+      responder.sendString(HttpResponseStatus.NOT_FOUND,
```
Comment (poornachandra) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AppLifecycleHttpHandler.java]: This too should be `NotFoundException`, right?

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/ProgramLifecycleHttpHandler.java [hunk 3]
```diff
@@ -437,27 +424,21 @@ public void saveProgramRuntimeArgs(HttpRequest request, HttpResponder responder,
...
-        return;
-      }
-      Map<String, String> args = decodeArguments(request);
-      preferencesStore.setProperties(namespaceId, appId, programType, programId, args);
-      responder.sendStatus(HttpResponseStatus.OK);
-    } catch (Throwable e) {
-      LOG.error("Error getting runtime args {}", e.getMessage(), e);
-      responder.sendStatus(HttpResponseStatus.INTERNAL_SERVER_ERROR);
+    if (!store.programExists(id, type)) {
+      responder.sendString(HttpResponseStatus.NOT_FOUND, "Program not found");
```
Comment (poornachandra) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/ProgramLifecycleHttpHandler.java]: `NotFoundException` here too.

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/ProgramLifecycleHttpHandler.java [hunk 4]
```diff
@@ -1626,7 +1602,12 @@ private AppFabricServiceStatus start(final Id.Program id, ProgramType type,
 
         @Override
         public void init(ProgramController.State state) {
-          store.setStart(id, runId, TimeUnit.SECONDS.convert(System.currentTimeMillis(), TimeUnit.MILLISECONDS));
+          try {
+            store.setStart(id, runId, TimeUnit.SECONDS.convert(System.currentTimeMillis(), TimeUnit.MILLISECONDS));
+          } catch (NotFoundException e) {
+            throw Throwables.propagate(e);
```
Comment (poornachandra) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/ProgramLifecycleHttpHandler.java]: Will the propagated exception get converted into `NotFoundException`?

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/ScheduleTaskRunner.java [hunk 5]
```diff
@@ -144,24 +153,33 @@ public void init(ProgramController.State state) {
...
-                  programId.getNamespaceId(), programId.getApplicationId(), programId.getId());
-        latch.countDown();
+        try {
+          store.setStop(programId, runId,
+                        TimeUnit.SECONDS.convert(System.currentTimeMillis(), TimeUnit.MILLISECONDS),
+                        ProgramController.State.STOPPED);
+        } catch (NotFoundException e) {
+          throw Throwables.propagate(e);
+        } finally {
+          LOG.debug("Program {} {} {} completed successfully.",
```
Comment (poornachandra) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/ScheduleTaskRunner.java]: If you move the log message before `store.setStop` then you won't need the finally block.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/ScheduleTaskRunner.java [hunk 6]
```diff
@@ -144,24 +153,33 @@ public void init(ProgramController.State state) {
...
-
-        latch.countDown();
+        try {
+          store.setStop(programId, runId,
+                        TimeUnit.SECONDS.convert(System.currentTimeMillis(), TimeUnit.MILLISECONDS),
+                        ProgramController.State.ERROR);
+        } catch (NotFoundException e) {
+          throw Throwables.propagate(e);
+        } finally {
+          LOG.debug("Program {} {} {} execution failed.",
```
Comment (poornachandra) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/ScheduleTaskRunner.java]: If you move the log message before `store.setStop` then you won't need the finally block.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java [hunk 7]
```diff
@@ -270,105 +304,212 @@ public ProgramArgs getProgramArgs(Id.Program program) {
...
+    protected void appendKey(MDSKey.Builder builder, Id.Adapter id) {
+      builder.add(id.getId());
+    }
   }
 
-  private MDSKey getNamespaceKey(@Nullable String name) {
-    MDSKey.Builder builder = new MDSKey.Builder().add(TYPE_NAMESPACE);
-    if (null != name) {
-      builder.add(name);
+//  private static final class ProgramView extends MetadataStoreView<Id.Application, Id.Program, ApplicationMeta> {
```
Comment (poornachandra) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java]: Commented out code.

**Comment by jwang47** [comment]
Closing to reopen for next release, as this doesn't add very much value but will introduce conflicts, risk, and will require an upgrade step.

**Comment by jwang47** [comment]
Also the approach for the `AppMetadataStore` refactoring is a big change, which should be discussed with some more people first.

**Pull request closed without merging**